AstSemantics.md

```diff
@@ -202,11 +202,15 @@ Out of bounds accesses trap.
 
 In the MVP, linear memory can be resized by a `grow_memory` operator. The
 operand to this operator is in units of the WebAssembly page size,
-which is 64KiB on all engines (though large page support may be added in 
+which is defined to be 64KiB (though large page support may be added in 
 the [future](FutureFeatures.md#large-page-support)).
 
  * `grow_memory` : grow linear memory by a given unsigned delta of pages.
-    Return the previous memory size in bytes.
+    Return the previous memory size in units of pages or -1 on failure.
+
+`grow_memory` must fail when attempting to grow past the maximum declared
+size, if one is specified in the [memory section](Module.md#linear-memory-section).
+`grow_memory` may also fail if an underlying system allocation fails.
```

Comment:
There's nothing special about `VirtualAlloc` here, `mmap`, `malloc` and `calloc` all do the same: this is just the behavior you'd expect on a OS which doesn't have a page file or runs out of page file space.  That means, in JS, if you `new ArrayBuffer(1024*1024*1024)`, since we're calling `calloc`, any JS element access can kill the process on such an OS.  For that matter any GC field access can too if it's in a fresh page allocated by the GC.  We could add a note to the Nondeterminism.md that this is another kind of resource exhaustion, but there's nothing specific to `grow_memory` here: to avoid this, we'd have to take extra steps (on both Windows and POSIX) in `grow_memory` to preemptively dirty pages which would be silly and noone would expect us to do.